File: code/ex274.java
class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;    if (n == 0)    return 0;
        Arrays.sort(citations);
        
        int h = 1;
        for (int i = n - 1; i >= 1; --i)
        {
            if (citations[i] >= h && citations[i - 1] <= h)
                break;
            
            h++;
        } 
        
        if (h == n && citations[0] < h)    h = 0;
        
        return h;
    }
}

File: code/ex274.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        sort(citations.begin(), citations.end());
        int lo = 0, hi = n - 1;
        while (lo <= hi) {
            int h = lo + (hi - lo) / 2;
            if (citations[h] == n - h)    return n - h;
            else if (citations[h] >= n - h)    hi = h - 1;
            else    lo = h + 1;
        }
        return n - lo;
    }
};